src/shared/pkcs11-util.c

```diff
@@ -691,10 +908,76 @@ int pkcs11_token_decrypt_data(
                 return log_error_errno(SYNTHETIC_ERRNO(EIO),
                                        "Failed to decrypt key on security token: %s", p11_kit_strerror(rv));
 
+        *ret_decrypted_data = dbuffer;
+        *ret_decrypted_data_size = dbuffer_size;
+        return 0;
+}
+
+/* We only care about asymmetric key so only asymmetric key type is listed here */
+static const char *pkcs11_key_type_to_string(CK_KEY_TYPE type) {
+        switch(type) {
+                case CKK_EC:
+                        return "ec";
+                case CKK_RSA:
+                        return "rsa";
+                case CKK_EC_EDWARDS:
+                        return "ed25519";
+                default:
+                        return NULL;
+        }
+}
+
+int pkcs11_token_decrypt_data(
+                CK_FUNCTION_LIST *m,
+                CK_SESSION_HANDLE session,
+                CK_OBJECT_HANDLE object,
+                const void *encrypted_data,
+                size_t encrypted_data_size,
+                void **ret_decrypted_data,
+                size_t *ret_decrypted_data_size) {
+
+        _cleanup_(erase_and_freep) void *decrypted_data = NULL;
+        size_t decrypted_data_size = 0;
+        CK_KEY_TYPE key_type;
+        CK_ATTRIBUTE key_type_template = {
+                .type = CKA_KEY_TYPE,
+                .pValue = &key_type,
+                .ulValueLen = sizeof(key_type)
+        };
+        CK_RV rv;
+        int r;
+
+        assert_se(m);
+        assert_se(encrypted_data);
+        assert_se(encrypted_data_size > 0);
+        assert_se(ret_decrypted_data);
+        assert_se(ret_decrypted_data_size);
+
+        rv = m->C_GetAttributeValue(session, object, &key_type_template, 1);
+        if (rv != CKR_OK)
+                return log_error_errno(EIO, "Failed to retrieve private key type");
+
+        if (key_type == CKK_RSA)
+                r = pkcs11_token_decrypt_data_rsa(m, session, object, encrypted_data, encrypted_data_size, &decrypted_data, &decrypted_data_size);
+#if HAVE_OPENSSL
+        else if (key_type == CKK_EC)
+                r = pkcs11_token_decrypt_data_ec(m, session, object, encrypted_data, encrypted_data_size, &decrypted_data, &decrypted_data_size);
+#endif
+        else {
+                const char *key_type_str = pkcs11_key_type_to_string(key_type);
+                if (key_type_str != NULL)
+                        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Unsupported private key type: %s", key_type_str);
+                else
+                        return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Unsupported private key type: %ld", key_type);
```

Comment:
we don't really know what type `key_type` has, CK_KEY_TYPE could be anything, hence maybe cast it to (unsigned long) here?